Slack is relying on garbage collection to stop camera capture

youennf opened the pull request:
#### 6f897d3266250fcbab3b622a0c71f1fd1dcba398
<pre>
Slack is relying on garbage collection to stop camera capture
<a href="https://rdar.apple.com/110627788">rdar://110627788</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288372">https://bugs.webkit.org/show_bug.cgi?id=288372</a>

Reviewed by Jean-Yves Avenard.

We allow GC of MediaStream and MediaStreamTrack objects when they do not have event listeners.
We also allow GC of non RTCPeerConnection MediaStream since they cannot fire events.

Drive-by fix: Remove an unneeded map in LibWebRTCMediaEndpoint.

Covered by added test.

* LayoutTests/fast/mediastream/mediastream-gc-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastream-gc.html: Added.
* Source/WebCore/Modules/mediastream/MediaStream.cpp:
(WebCore::MediaStream::create):
(WebCore::MediaStream::MediaStream):
(WebCore::MediaStream::didAddTrack):
(WebCore::MediaStream::didRemoveTrack):
(WebCore::MediaStream::addTrackFromPlatform):
(WebCore::MediaStream::inactivate):
(WebCore::MediaStream::virtualHasPendingActivity const):
(WebCore::MediaStream::stop): Deleted.
* Source/WebCore/Modules/mediastream/MediaStream.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::~MediaStreamTrack):
(WebCore::MediaStreamTrack::virtualHasPendingActivity const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::mediaStreamFromRTCStream):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::mediaStreamFromRTCStreamId):
(WebCore::LibWebRTCMediaEndpoint::stop):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::audioCaptureSourceCount const):
(WebCore::PlatformMediaSessionManager::hasAudioCaptureSource const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::removeMediaStreamTrack):
(WebCore::Internals::audioCaptureSourceCount const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/291110@main">https://commits.webkit.org/291110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9b8e28e6293a7814a88d2bdab23bb88642a03a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1135 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96965 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42627 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94096 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20072 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70618 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28091 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95047 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83371 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50947 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/978 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41843 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/965 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99005 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19155 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19407 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79229 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19524 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23417 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12182 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19136 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18831 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStream.cpp:
```diff
@@ -369,14 +374,14 @@ Document* MediaStream::document() const
 }
 
-void MediaStream::stop()
+void MediaStream::inactivate()
```
considering this is always called by `stop()` and this new `inactivate` is called by `void LibWebRTCMediaEndpoint::stop()` shouldn't we leave it as is ?

youennf replied:
stop is ActiveDOMObject::stop. I think it is clearer this way.

jyavenard commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
```diff
@@ -192,7 +192,6 @@ class LibWebRTCMediaEndpoint
     SetRemoteSessionDescriptionObserver<LibWebRTCMediaEndpoint> m_setRemoteSessionDescriptionObserver;
 
     MemoryCompactRobinHoodHashMap<String, RefPtr<MediaStream>> m_remoteStreamsById;
-    HashMap<MediaStreamTrack*, Vector<String>> m_remoteStreamsFromRemoteTrack;
```
should have a note in the changelog about removing an unused member

youennf replied:
Added

rniwa commented on LayoutTests/fast/mediastream/mediastream-gc.html:
```diff
@@ -0,0 +1,49 @@
+}, "Media streams can be gced if they are not expected to fire events - 1");
+
+promise_test(async () => {
+    video.srcObject = await navigator.mediaDevices.getUserMedia({audio:true, video : true});
```
This test is going to be flaky because we have a conservative GC.
Retry 5-10 times and check that at least one of them gets collected instead.

youennf committed:
Slack is relying on garbage collection to stop camera capture
rdar://110627788
https://bugs.webkit.org/show_bug.cgi?id=288372

Reviewed by Jean-Yves Avenard.

We allow GC of MediaStream and MediaStreamTrack objects when they do not have event listeners.
We also allow GC of non RTCPeerConnection MediaStream since they cannot fire events.

Drive-by fix: Remove an unneeded map in LibWebRTCMediaEndpoint.

Covered by added test.

* LayoutTests/fast/mediastream/mediastream-gc-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastream-gc.html: Added.
* Source/WebCore/Modules/mediastream/MediaStream.cpp:
(WebCore::MediaStream::create):
(WebCore::MediaStream::MediaStream):
(WebCore::MediaStream::didAddTrack):
(WebCore::MediaStream::didRemoveTrack):
(WebCore::MediaStream::addTrackFromPlatform):
(WebCore::MediaStream::inactivate):
(WebCore::MediaStream::virtualHasPendingActivity const):
(WebCore::MediaStream::stop): Deleted.
* Source/WebCore/Modules/mediastream/MediaStream.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::~MediaStreamTrack):
(WebCore::MediaStreamTrack::virtualHasPendingActivity const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::mediaStreamFromRTCStream):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::mediaStreamFromRTCStreamId):
(WebCore::LibWebRTCMediaEndpoint::stop):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::audioCaptureSourceCount const):
(WebCore::PlatformMediaSessionManager::hasAudioCaptureSource const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::removeMediaStreamTrack):
(WebCore::Internals::audioCaptureSourceCount const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/291110@main
